File: src/layout.ts
import cytoscape from "cytoscape";
import type { NetPosition } from "./model";

export type LayoutNode = { id: string; width: number; height: number };
export type LayoutEdge = { source: string; target: string };

export type LayoutOptions = {
  /** Force-directed repulsion strength (cose). */
  nodeRepulsion?: number;
  /** Force-directed ideal edge length (cose). */
  idealEdgeLength?: number;
  /** Concentric ring spacing (concentric). */
  ringSpacing?: number;
};

/**
 * Scale force-directed coordinates so the minimum node gap reaches `gap` (small
 * graphs), or fit the bounding box to a target area (large graphs).
 */
function normalizeToGap(
  centers: Record<string, { x: number; y: number }>,
  gap: number,
): Record<string, { x: number; y: number }> {
  const ids = Object.keys(centers);
  if (ids.length === 0) return centers;
  let minX = Infinity;
  let minY = Infinity;
  for (const id of ids) {
    minX = Math.min(minX, centers[id].x);
    minY = Math.min(minY, centers[id].y);
  }
  const pts = ids.map((id) => ({ id, x: centers[id].x - minX, y: centers[id].y - minY }));
  let scale = 1;
  if (pts.length <= 150) {
    let minGap = Infinity;
    for (let i = 0; i < pts.length; i++) {
      for (let j = i + 1; j < pts.length; j++) {
        const d = Math.hypot(pts[i].x - pts[j].x, pts[i].y - pts[j].y);
        if (d > 0 && d < minGap) minGap = d;
      }
    }
    if (minGap > 0 && minGap < gap) scale = gap / minGap;
  } else {
    let maxX = 0;
    let maxY = 0;
    for (const p of pts) {
      maxX = Math.max(maxX, p.x);
      maxY = Math.max(maxY, p.y);
    }
    scale = Math.min(1600 / (maxX || 1), 1200 / (maxY || 1));
  }
  const out: Record<string, { x: number; y: number }> = {};
  for (const p of pts) out[p.id] = { x: p.x * scale, y: p.y * scale };
  return out;
}

function toTopLeft(
  centers: Record<string, { x: number; y: number }>,
  nodes: LayoutNode[],
): Record<string, NetPosition> {
  const out: Record<string, NetPosition> = {};
  for (const n of nodes) {
    const c = centers[n.id];
    if (c) out[n.id] = { x: c.x - n.width / 2, y: c.y - n.height / 2 };
  }
  return out;
}

function cyElements(nodes: LayoutNode[], edges: LayoutEdge[]) {
  return [
    ...nodes.map((n) => ({ data: { id: n.id } })),
    ...edges.map((e, i) => ({ data: { id: `le${i}`, source: e.source, target: e.target } })),
  ];
}

/**
 * Force-directed layout (Cytoscape `cose`). Keeps connected nodes close, so
 * arcs stay short even for cyclic nets — unlike layered (dagre) layouts that
 * stretch back-edges across the whole canvas.
 */
export function computeLayout(
  nodes: LayoutNode[],
  edges: LayoutEdge[],
  opts: LayoutOptions = {},
): Record<string, NetPosition> {
  if (nodes.length === 0) return {};
  const cy = cytoscape({
    headless: true,
    elements: cyElements(nodes, edges),
  });
  cy.layout({
    name: "cose",
    animate: false,
    nodeRepulsion: () => opts.nodeRepulsion ?? 60000,
    idealEdgeLength: () => opts.idealEdgeLength ?? 260,
    nodeOverlap: 40,
    edgeElasticity: () => 60,
  }).run();
  const centers: Record<string, { x: number; y: number }> = {};
  cy.nodes().forEach((n) => {
    const p = n.position();
    centers[n.id()] = { x: p.x, y: p.y };
  });
  return toTopLeft(normalizeToGap(centers, 130), nodes);
}

/**
 * Concentric layout by a per-node level (e.g. BFS distance from the initial
 * state). The level-0 node sits at the center and rings grow outward, so
 * back-edges between adjacent rings stay short.
 */
export function computeConcentricLayout(
  nodes: LayoutNode[],
  edges: LayoutEdge[],
  level: Record<string, number>,
  opts: LayoutOptions = {},
): Record<string, NetPosition> {
  if (nodes.length === 0) return {};
  const cy = cytoscape({
    headless: true,
    elements: cyElements(nodes, edges),
  });
  cy.layout({
    name: "concentric",
    animate: false,
    concentric: (n) => level[n.id()] ?? 0,
    levelWidth: () => opts.ringSpacing ?? 120,
    minNodeSpacing: opts.ringSpacing ?? 120,
    startAngle: -Math.PI / 2,
    clockwise: true,
    equidistant: true,
  }).run();
  const centers: Record<string, { x: number; y: number }> = {};
  cy.nodes().forEach((n) => {
    const p = n.position();
    centers[n.id()] = { x: p.x, y: p.y };
  });
  return toTopLeft(normalizeToGap(centers, 130), nodes);
}

File: dioxus-app/src/layout.rs
use std::collections::HashMap;

use crate::model::{Edge, Node, Position};

/// Simple force-directed layout (Fruchterman–Reingold style) that keeps
/// connected nodes close. Replaces the JS Cytoscape "cose" layout.
pub fn compute_layout(nodes: &[Node], edges: &[Edge], opts: &LayoutOptions) -> HashMap<String, Position> {
    let mut pos: HashMap<String, Position> = HashMap::new();
    for n in nodes {
        pos.insert(n.id.clone(), Position { x: n.position.x, y: n.position.y });
    }
    if nodes.is_empty() {
        return pos;
    }

    let area = opts.area;
    let k = (area / nodes.len().max(1) as f32).sqrt();
    let repulsion = opts.repulsion;
    let ideal_len = opts.ideal_edge_length;

    let mut velocity: HashMap<String, (f32, f32)> = HashMap::new();
    for n in nodes {
        velocity.insert(n.id.clone(), (0.0, 0.0));
    }

    for _iter in 0..opts.iterations {
        let mut disp: HashMap<String, (f32, f32)> = HashMap::new();
        for n in nodes {
            disp.insert(n.id.clone(), (0.0, 0.0));
        }

        // repulsive forces
        for i in 0..nodes.len() {
            for j in (i + 1)..nodes.len() {
                let a = &nodes[i];
                let b = &nodes[j];
                let (dx, dy) = delta(&pos, &a.id, &b.id);
                let d2 = (dx * dx + dy * dy).max(0.01);
                let d = d2.sqrt();
                let force = repulsion / d2;
                let fx = dx / d * force;
                let fy = dy / d * force;
                disp.entry(a.id.clone()).or_default().0 -= fx;
                disp.entry(a.id.clone()).or_default().1 -= fy;
                disp.entry(b.id.clone()).or_default().0 += fx;
                disp.entry(b.id.clone()).or_default().1 += fy;
            }
        }

        // attractive forces along edges
        for e in edges {
            let (dx, dy) = delta(&pos, &e.source, &e.target);
            let d = (dx * dx + dy * dy).sqrt().max(0.01);
            let force = (d * d) / (k * ideal_len);
            let fx = dx / d * force;
            let fy = dy / d * force;
            disp.entry(e.source.clone()).or_default().0 += fx;
            disp.entry(e.source.clone()).or_default().1 += fy;
            disp.entry(e.target.clone()).or_default().0 -= fx;
            disp.entry(e.target.clone()).or_default().1 -= fy;
        }

        // apply with temperature cooling
        let temp = opts.temperature * (1.0 - _iter as f32 / opts.iterations as f32);
        for n in nodes {
            let (dx, dy) = disp[&n.id];
            let d = (dx * dx + dy * dy).sqrt().max(0.01);
            let mx = dx / d * temp.min(d);
            let my = dy / d * temp.min(d);
            let v = velocity.entry(n.id.clone()).or_insert((0.0, 0.0));
            v.0 = v.0 * 0.85 + mx * 0.15;
            v.1 = v.1 * 0.85 + my * 0.15;
            if let Some(p) = pos.get_mut(&n.id) {
                p.x += v.0;
                p.y += v.1;
            }
        }
    }

    normalize_to_gap(&mut pos, 130.0);
    pos
}

fn delta(pos: &HashMap<String, Position>, a: &str, b: &str) -> (f32, f32) {
    let pa = pos.get(a).cloned().unwrap_or(Position { x: 0.0, y: 0.0 });
    let pb = pos.get(b).cloned().unwrap_or(Position { x: 0.0, y: 0.0 });
    (pa.x - pb.x, pa.y - pb.y)
}

fn normalize_to_gap(pos: &mut HashMap<String, Position>, gap: f32) {
    if pos.is_empty() {
        return;
    }
    let min_x = pos.values().map(|p| p.x).fold(f32::INFINITY, f32::min);
    let min_y = pos.values().map(|p| p.y).fold(f32::INFINITY, f32::min);
    for p in pos.values_mut() {
        p.x -= min_x;
        p.y -= min_y;
    }
    let ids: Vec<String> = pos.keys().cloned().collect();
    let mut min_gap = f32::INFINITY;
    for i in 0..ids.len() {
        for j in (i + 1)..ids.len() {
            let d = dist(&pos[&ids[i]], &pos[&ids[j]]);
            if d > 0.0 && d < min_gap {
                min_gap = d;
            }
        }
    }
    if min_gap > 0.0 && min_gap < gap {
        let scale = gap / min_gap;
        for p in pos.values_mut() {
            p.x *= scale;
            p.y *= scale;
        }
    }
}

fn dist(a: &Position, b: &Position) -> f32 {
    ((a.x - b.x) * (a.x - b.x) + (a.y - b.y) * (a.y - b.y)).sqrt()
}

pub struct LayoutOptions {
    pub iterations: usize,
    pub temperature: f32,
    pub repulsion: f32,
    pub ideal_edge_length: f32,
    pub area: f32,
}

impl Default for LayoutOptions {
    fn default() -> Self {
        LayoutOptions {
            iterations: 200,
            temperature: 400.0,
            repulsion: 60000.0,
            ideal_edge_length: 260.0,
            area: 400000.0,
        }
    }
}